pkg/frontend/openshiftcluster_preflightvalidation.go

```diff
@@ -0,0 +1,125 @@
+		ch := make(chan api.ValidationResult, resourceCount)
+		for _, raw := range resources.Resources {
+			// get typeMeta from the raw data
+			typeMeta := api.ResourceTypeMeta{}
+			if err := json.Unmarshal(raw, &typeMeta); err != nil {
+				// failing to parse the preflight body is not considered a validation failure. continue
+				log.Warningf("bad request. Failed to unmarshal ResourceTypeMeta: %s", err)
+				continue
+			}
+			if strings.EqualFold(typeMeta.Type, "Microsoft.RedHatOpenShift/openShiftClusters") {
+				ch <- f._preflightValidation(ctx, log, raw, typeMeta.APIVersion)
+			}
+		}
+		close(ch)
+		// define the input, and what expected output will be. Better off to write my own, instead of using existing UT
+		for res := range ch {
+			// go through and return the first error
+			if res.Status == api.ValidationStatusFailed {
+				log.Warningf("preflight validation failed")
+				b = marshalValidationResult(res)
+				reply(log, w, header, b, statusCodeError(http.StatusOK))
+				return
+			}
+			b = marshalValidationResult(res)
+		}
+		log.Info("preflight validation succeeded")
+		reply(log, w, header, b, statusCodeError(http.StatusOK))
+	} else {
+		b = marshalValidationResult(api.ValidationResult{
+			Status: api.ValidationStatusFailed,
+			Error: &api.ManagementErrorWithDetails{
+				Message: to.StringPtr(err.Error()),
+			},
+		})
+		reply(log, w, header, b, statusCodeError(http.StatusOK))
+	}
+}
+
+func (f *frontend) _preflightValidation(ctx context.Context, log *logrus.Entry, raw json.RawMessage, apiVersion string) api.ValidationResult {
+	// unmarshal raw to OpenShiftCluster type
+	doc := &api.OpenShiftCluster{}
+	doc.Properties.ProvisioningState = api.ProvisioningStateSucceeded
+
+	if !f.env.IsLocalDevelopmentMode() /* not local dev or CI */ {
+		doc.Properties.FeatureProfile.GatewayEnabled = true
+	}
+
+	converter := f.apis[apiVersion].OpenShiftClusterConverter
+	ext := converter.ToExternal(doc)
+	if err := json.Unmarshal(raw, &ext); err != nil {
+		return api.ValidationResult{
+			Status: api.ValidationStatusFailed,
+			Error: &api.ManagementErrorWithDetails{
+				Message: to.StringPtr(err.Error()),
+			},
+		}
+	}
+
+	converter.ToInternal(ext, doc)
+
+	if err := f.validateInstallVersion(ctx, doc); err != nil {
+		return api.ValidationResult{
+			Status: api.ValidationStatusFailed,
+			Error: &api.ManagementErrorWithDetails{
+				Code:    to.StringPtr(api.CloudErrorCodeInvalidParameter),
+				Message: to.StringPtr(err.Error()),
+			},
+		}
+	}
+
+	return validationSuccess
+}
+
+func unmarshalRequest(body []byte) (*api.PreflightRequest, error) {
+	preflightRequest := &api.PreflightRequest{}
+	if err := json.Unmarshal(body, preflightRequest); err != nil {
+		return nil, api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized: %q.", err)
+	}
+	return preflightRequest, nil
+}
+
+var validationSuccess = api.ValidationResult{
```

Comment:
can you move this var before the functions, which is a more general go convention